airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumMessageProducer.java

```diff
@@ -0,0 +1,170 @@
+    this.initialOffset = new HashMap<>(this.previousCheckpointOffset);
+    resetCheckpointValues();
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage generateStateMessageAtCheckpoint(ConfiguredAirbyteStream stream) {
+    LOGGER.info("Sending CDC checkpoint state message.");
+    final AirbyteStateMessage stateMessage = createStateMessage(checkpointOffsetToSend);
+    previousCheckpointOffset.clear();
+    previousCheckpointOffset.putAll(checkpointOffsetToSend);
+    sendCheckpointMessage = false;
+    return stateMessage;
+  }
+
+  /**
+   * @param stream
+   * @param message
+   * @return
+   */
+  @Override
+  public AirbyteMessage processRecordMessage(ConfiguredAirbyteStream stream, ChangeEventWithMetadata message) {
+
+    if (checkpointOffsetToSend.isEmpty()) {
+      try {
+        final HashMap<String, String> temporalOffset = (HashMap<String, String>) offsetManager.read();
+        if (!targetPosition.isSameOffset(previousCheckpointOffset, temporalOffset)) {
+          checkpointOffsetToSend.putAll(temporalOffset);
+        }
+      } catch (final ConnectException e) {
+        LOGGER.warn("Offset file is being written by Debezium. Skipping CDC checkpoint in this loop.");
+      }
+    }
+
+    if (checkpointOffsetToSend.size() == 1
+        && !message.isSnapshotEvent()) {
+      if (targetPosition.isEventAheadOffset(checkpointOffsetToSend, message)) {
+        sendCheckpointMessage = true;
+      } else {
+        LOGGER.info("Encountered records with the same event offset.");
+      }
+    }
+
+    return eventConverter.toAirbyteMessage(message);
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage createFinalStateMessage(ConfiguredAirbyteStream stream) {
+    final var syncFinishedOffset = (HashMap<String, String>) offsetManager.read();
+
+    System.out.println("initialoffset: " + initialOffset);
+    System.out.println("syncFinishedOffset: " + syncFinishedOffset);
+
+    if (targetPosition.isSameOffset(initialOffset, syncFinishedOffset)) {
+      // Edge case where no progress has been made: wrap up the
+      // sync by returning the initial offset instead of the
+      // current offset. We do this because we found that
+      // for some databases, heartbeats will cause Debezium to
+      // overwrite the offset file with a state which doesn't
+      // include all necessary data such as snapshot completion.
+      // This is the case for MS SQL Server, at least.
+      return createStateMessage(initialOffset);
+    }
+    return createStateMessage(syncFinishedOffset);
+  }
+
+  private void resetCheckpointValues() {
+    checkpointOffsetToSend.clear();
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public boolean shouldEmitStateMessage(ConfiguredAirbyteStream stream) {
+    return cdcStateHandler.isCdcCheckpointEnabled() && sendCheckpointMessage;
```

Comment:
I thought that we always want `cdcStateHandler.isCdcCheckpointEnabled()` to return true. I suggest we remove this condition and simply rename the `sendCheckpointMessage` variable to `shouldEmitStateMessage` and do without this method.

Reply:
+1 - all our cdc iterators support checkpointing, so let's get rid of this